File: aegis_science_demo/handshake_oops_engine.py
#!/usr/bin/env python3
"""
ROOPS (Robot Object-Oriented Programming System - OOPS)
Python High-Performance Sub-Object Modular Architecture for Humanoid Robot Hands
Author: Aegis (Google DeepMind Science / ROOPS Infrastructure Hub)
Dual C++ & Python High-Performance Engine Implementation
"""

import math
import time
import json

class Vector3:
    def __init__(self, x=0.0, y=0.0, z=0.0):
        self.x = x
        self.y = y
        self.z = z

class EulerRotation:
    def __init__(self, rx=0.0, ry=0.0, rz=0.0):
        self.rx = rx
        self.ry = ry
        self.rz = rz

class TcpFrameAxes:
    def __init__(self, length=3.5):
        self.axis_length = length

class CollisionShield:
    def __init__(self, hex_color="0x34d399", opacity=0.7):
        self.color_hex = hex_color
        self.opacity = opacity

    def set_color(self, hex_color):
        self.color_hex = hex_color

class FingerJointLink:
    def __init__(self, name, x, z, length, radius, max_angle):
        self.name = name
        self.x = x
        self.z = z
        self.length = length
        self.radius = radius
        self.max_angle = max_angle
        self.current_curl_angle = 0.0
        self.shield = CollisionShield("0x34d399", 0.7)

    def set_curl_angle(self, angle):
        self.current_curl_angle = min(angle, self.max_angle)

class PalmBase:
    def __init__(self, width=3.6, height=1.2, depth=2.4):
        self.width = width
        self.height = height
        self.depth = depth
        self.shield = CollisionShield("0x34d399", 0.6)

class RobotHandObject:
    def __init__(self, name, is_facing_rotated=False):
        self.name = name
        self.is_facing_rotated = is_facing_rotated
        self.palm = PalmBase(3.6, 1.2, 2.4)
        self.tcp_axes = TcpFrameAxes(3.5)
        
        self.internal_state = {
            "handName": name,
            "px": 0.25 if is_facing_rotated else -0.25,
            "py": 3.0,
            "pz": 4.0 if is_facing_rotated else -4.0,
            "rx": 0.0,
            "ry": math.pi if is_facing_rotated else 0.0,
            "rz": 0.0,
            "lastUpdated": int(time.time() * 1000)
        }

        self.fingers = [
            FingerJointLink("thumb", -1.6, 1.0, 2.2, 0.35, 1.10),
            FingerJointLink("index", -0.8, 1.4, 2.8, 0.32, 1.22),
            FingerJointLink("middle", 0.0, 1.5, 3.0, 0.32, 1.25),
            FingerJointLink("ring", 0.8, 1.4, 2.7, 0.32, 1.20),
            FingerJointLink("pinky", 1.6, 1.2, 2.2, 0.30, 1.15)
        ]

    def set_position(self, x, y, z):
        self.internal_state["px"] = x
        self.internal_state["py"] = y
        self.internal_state["pz"] = z

    def set_rotation(self, rx, ry, rz):
        self.internal_state["rx"] = rx
        self.internal_state["ry"] = ry
        self.internal_state["rz"] = rz

    def set_finger_curl(self, curl_amount):
        for f in self.fingers:
            f.set_curl_angle(curl_amount)

    def set_all_shield_colors(self, hex_color):
        self.palm.shield.set_color(hex_color)
        for f in self.fingers:
            f.shield.set_color(hex_color)

    def export_to_external_json(self, filepath):
        with open(filepath, "w") as f:
            json.dump(self.internal_state, f, indent=2)
        print(f"[ROOPS OOPS Python] Exported {self.name} state to {filepath}")

    def print_telemetry(self):
        st = self.internal_state
        print(f"[{self.name}] Pos: ({st['px']:.2f}, {st['py']:.2f}, {st['pz']:.2f}) | "
              f"Rot: ({st['rx']*180/math.pi:.1f}deg, {st['ry']*180/math.pi:.1f}deg, {st['rz']*180/math.pi:.1f}deg) | "
              f"Shield: {self.palm.shield.color_hex}")

def main():
    print("=========================================================")
    print("  ROOPS OOPS (Object-Oriented Python) Humanoid Hand Engine  ")
    print("=========================================================")

    handA = RobotHandObject("Hand A", False)
    handB = RobotHandObject("Hand B", True)

    handA.print_telemetry()
    handB.print_telemetry()

    # Simulate Sequential Trajectory
    for step in range(6):
        slider_val = step * 0.05
        approach_ratio = min(slider_val / 0.185, 1.0)
        clasp_ratio = max((slider_val - 0.185) / 0.115, 0.0)

        base_z = (1.0 - approach_ratio) * 3.5 + 1.35
        base_curl = approach_ratio * 0.35 + clasp_ratio * 0.90

        handA.set_position(-0.25, 3.0, -base_z)
        handB.set_position(0.25, 3.0, base_z)

        handA.set_finger_curl(base_curl)
        handB.set_finger_curl(base_curl)

        if approach_ratio >= 1.0:
            handA.set_all_shield_colors("0x34d399")
            handB.set_all_shield_colors("0x34d399")
        else:
            handA.set_all_shield_colors("0x38bdf8")
            handB.set_all_shield_colors("0xf97316")

        print(f"\n--- Step {step} (Slider: {slider_val:.2f}m) ---")
        handA.print_telemetry()
        handB.print_telemetry()

    handA.export_to_external_json("/home/moos/dev_ws/dual_arms/aegis_science_demo/tcp_handA_py.json")
    handB.export_to_external_json("/home/moos/dev_ws/dual_arms/aegis_science_demo/tcp_handB_py.json")
    print("\n[ROOPS Python Engine Execution Completed Successfully 🟢]")

if __name__ == "__main__":
    main()

File: aegis_science_demo/handshake_oops_engine.cpp
/**
 * ROOPS (Robot Object-Oriented Programming System - OOPS)
 * C++ High-Performance Sub-Object Modular Architecture for Humanoid Robot Hands
 * Author: Aegis (Google DeepMind Science / ROOPS Infrastructure Hub)
 * Dual C++ & Python High-Performance Engine Implementation
 */

#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <fstream>
#include <memory>
#include <chrono>

namespace ROOPS {

// 3D Vector Math
struct Vector3 {
    double x = 0.0, y = 0.0, z = 0.0;
    Vector3() = default;
    Vector3(double _x, double _y, double _z) : x(_x), y(_y), z(_z) {}
};

struct EulerRotation {
    double rx = 0.0, ry = 0.0, rz = 0.0; // In radians
    EulerRotation() = default;
    EulerRotation(double _rx, double _ry, double _rz) : rx(_rx), ry(_ry), rz(_rz) {}
};

// Sub-Object 1: TCP Frame Axes
class TcpFrameAxes {
private:
    double axisLength;
public:
    explicit TcpFrameAxes(double length = 3.5) : axisLength(length) {}
    void RenderInfo() const {
        std::cout << "  [TcpFrameAxes] RGB Axes active (Length: " << axisLength << "m)\n";
    }
};

// Sub-Object 2: 3D Collision Shield
class CollisionShield {
private:
    std::string colorHex;
    double opacity;
    bool active;
public:
    CollisionShield(std::string hex = "0x34d399", double op = 0.7)
        : colorHex(hex), opacity(op), active(true) {}
    void SetColor(const std::string& hex) { colorHex = hex; }
    std::string GetColor() const { return colorHex; }
};

// Sub-Object 3: Finger Joint Link Sub-Object
class FingerJointLink {
public:
    std::string name;
    double x, z;
    double length;
    double radius;
    double maxAngle;
    double currentCurlAngle;
    CollisionShield shield;

    FingerJointLink(std::string _name, double _x, double _z, double _len, double _r, double _maxAngle)
        : name(_name), x(_x), z(_z), length(_len), radius(_r), maxAngle(_maxAngle), currentCurlAngle(0.0), shield("0x34d399", 0.7) {}

    void SetCurlAngle(double angle) {
        currentCurlAngle = std::min(angle, maxAngle);
    }
};

// Sub-Object 4: Palm Base Sub-Object
class PalmBase {
public:
    double width, height, depth;
    CollisionShield shield;

    PalmBase(double w = 3.6, double h = 1.2, double d = 2.4)
        : width(w), height(h), depth(d), shield("0x34d399", 0.6) {}
};

// Internal Instance State Struct
struct HandInternalState {
    std::string handName;
    Vector3 position;
    EulerRotation rotation;
    uint64_t lastUpdatedTimestamp;
};

// Sub-Object 5: Top-Level RobotHandObject
class RobotHandObject {
private:
    std::string name;
    bool isFacingRotated;
    PalmBase palm;
    TcpFrameAxes tcpAxes;
    std::vector<FingerJointLink> fingers;
    HandInternalState internalState;
    bool isSelected;

public:
    RobotHandObject(std::string _name, bool facingRotated = false)
        : name(_name), isFacingRotated(facingRotated), palm(3.6, 1.2, 2.4), tcpAxes(3.5), isSelected(false) {
        
        // Initial Pose
        internalState.handName = _name;
        internalState.position = facingRotated ? Vector3(0.25, 3.0, 4.0) : Vector3(-0.25, 3.0, -4.0);
        internalState.rotation = facingRotated ? EulerRotation(0.0, M_PI, 0.0) : EulerRotation(0.0, 0.0, 0.0);
        internalState.lastUpdatedTimestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count();

        // 5 Finger Joint Links initialization
        fingers.emplace_back("thumb", -1.6, 1.0, 2.2, 0.35, 1.10);
        fingers.emplace_back("index", -0.8, 1.4, 2.8, 0.32, 1.22);
        fingers.emplace_back("middle", 0.0, 1.5, 3.0, 0.32, 1.25);
        fingers.emplace_back("ring", 0.8, 1.4, 2.7, 0.32, 1.20);
        fingers.emplace_back("pinky", 1.6, 1.2, 2.2, 0.30, 1.15);
    }

    void SetPosition(double x, double y, double z) {
        internalState.position = Vector3(x, y, z);
    }

    void SetRotation(double rx, double ry, double rz) {
        internalState.rotation = EulerRotation(rx, ry, rz);
    }

    void SetFingerCurl(double curlAmount) {
        for (auto& finger : fingers) {
            finger.SetCurlAngle(curlAmount);
        }
    }

    void SetAllShieldColors(const std::string& hex) {
        palm.shield.SetColor(hex);
        for (auto& finger : fingers) {
            finger.shield.SetColor(hex);
        }
    }

    // Export internal state to JSON File for external Python/WebGL linking
    void ExportToExternalJson(const std::string& filepath) const {
        std::ofstream file(filepath);
        if (file.is_open()) {
            file << "{\n";
            file << "  \"handName\": \"" << internalState.handName << "\",\n";
            file << "  \"px\": " << internalState.position.x << ",\n";
            file << "  \"py\": " << internalState.position.y << ",\n";
            file << "  \"pz\": " << internalState.position.z << ",\n";
            file << "  \"rx\": " << internalState.rotation.rx << ",\n";
            file << "  \"ry\": " << internalState.rotation.ry << ",\n";
            file << "  \"rz\": " << internalState.rotation.rz << "\n";
            file << "}\n";
            file.close();
            std::cout << "[ROOPS OOPS C++] Exported " << name << " state to " << filepath << "\n";
        }
    }

    void PrintTelemetry() const {
        std::cout << "[" << name << "] Pos: (" << internalState.position.x << ", " 
                  << internalState.position.y << ", " << internalState.position.z << ") | Rot: ("
                  << internalState.rotation.rx * 180.0 / M_PI << "deg, "
                  << internalState.rotation.ry * 180.0 / M_PI << "deg, "
                  << internalState.rotation.rz * 180.0 / M_PI << "deg) | Shield: "
                  << palm.shield.GetColor() << "\n";
    }
};

} // namespace ROOPS

int main() {
    std::cout << "=========================================================\n";
    std::cout << "  ROOPS OOPS (Object-Oriented C++) Humanoid Hand Engine  \n";
    std::cout << "=========================================================\n";

    ROOPS::RobotHandObject handA("Hand A", false);
    ROOPS::RobotHandObject handB("Hand B", true);

    std::cout << "\n[C++ Engine Init Poses]\n";
    handA.PrintTelemetry();
    handB.PrintTelemetry();

    // Simulate Sequential Phase-Coupled Handshake Trajectory (Phase 1 & Phase 2)
    std::cout << "\n[Simulating Sequential Phase-Coupled Handshake Trajectory in C++]\n";
    for (int step = 0; step <= 5; ++step) {
        double sliderVal = step * 0.05; // 0.00m to 0.25m
        double approachRatio = std::min(sliderVal / 0.185, 1.0);
        double claspRatio = std::max((sliderVal - 0.185) / 0.115, 0.0);

        double baseZ = (1.0 - approachRatio) * 3.5 + 1.35;
        double baseCurl = approachRatio * 0.35 + claspRatio * 0.90;

        handA.SetPosition(-0.25, 3.0, -baseZ);
        handB.SetPosition(0.25, 3.0, baseZ);

        handA.SetFingerCurl(baseCurl);
        handB.SetFingerCurl(baseCurl);

        if (approachRatio >= 1.0) {
            handA.SetAllShieldColors("0x34d399"); // Green shields in Phase 2
            handB.SetAllShieldColors("0x34d399");
        } else {
            handA.SetAllShieldColors("0x38bdf8"); // Cyan / Orange in Phase 1
            handB.SetAllShieldColors("0xf97316");
        }

        std::cout << "\n--- Step " << step << " (Slider: " << sliderVal << "m) ---\n";
        handA.PrintTelemetry();
        handB.PrintTelemetry();
    }

    // Export Linked JSON
    handA.ExportToExternalJson("/home/moos/dev_ws/dual_arms/aegis_science_demo/tcp_handA_cpp.json");
    handB.ExportToExternalJson("/home/moos/dev_ws/dual_arms/aegis_science_demo/tcp_handB_cpp.json");

    std::cout << "\n[ROOPS C++ Engine Execution Completed Successfully 🟢]\n";
    return 0;
}
